### PR TITLE
Property declaration block serialization should check for importance

### LINKED ISSUE #14329: Property declaration block serialization should check for importance
See https://drafts.csswg.org/cssom/#serialize-a-css-declaration-block (step 3.3.2.4):

> If there is one or more CSS declarations in current longhands have their important flag set and one or more with it unset, continue with the steps labeled shorthand loop.

Just creating it since it doesn't seem to be tracked. There's a `TODO` in declaration_block.rs line 99 in this regard.

cc @Manishearth @SimonSapin and @canaltinova (who could be interested in working on this)

I don't have a clear estimate of how easy or hard this can be. Probably not too much, but should be backed with tests.

### PR DESCRIPTION
**Pull request by canova**
r? @emilio , @Manishearth  

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14329 (github issue number if applicable).

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14365)

### CODE REVIEW

**Comment by emilio** [comment]
@canaltinova what you should be testing is that something like:

```html
<div id="foo2" style="margin-right: 10px; margin-left: 10px; margin-top: 10px; margin-bottom: 10px;">foo</div>
```

Gets serialized as `margin: 10px`.

meanwhile:

```html
<div id="foo2" style="margin-right: 10px; margin-left: 10px; margin-top: 10px; margin-bottom: 10px!important;">foo</div>
```

Gets serialized as `margin-right: 10px; margin-left: 10px; margin-top: 10px; margin-bottom: 10px !important;`.

At least that's what I was intending to fill.

**Comment by canaltinova** [comment]
Oh, I misunderstood the problem. Sorry about that, I'll update the code soon.

**Comment by canaltinova** [comment]
@emilio It looks like your code examples are working as expected with `element.style.cssText`even before my patch.
But your second example shouldn't serialize with `element.style.margin` (that's also what other browsers do), we are currently serializing it like `margin: 10px`.
I couldn't find note about this in spec but it seems reasonable because when you add !important to a longhand it doesn't become a shorthand anymore(as your second example).
I'll update the code and tests to ensure that. But I'm open to other ideas :)

**Comment by canaltinova** [comment]
Updated the code.

File: components/style/properties/declaration_block.rs [hunk 1]
```diff
@@ -89,13 +89,21 @@ impl PropertyDeclarationBlock {
 
                 // Step 2.2.2 & 2.2.3
                 match declaration {
-                    Some(&(ref declaration, _importance)) => list.push(declaration),
+                    Some(&(ref declaration, importance)) => {
+                        // If there is an longhand with important, we don't
+                        // serialize it as a shorthand.
+                        if importance.important() {
```
Comment (emilio) [inline components/style/properties/declaration_block.rs]: We should still serialize it with `!important` if all of them have `!important`, so this is not enough.

Please add a test for it.

**Comment by canaltinova** [comment]
Updated code and added a test for it. Thanks!

File: components/style/properties/declaration_block.rs [hunk 2]
```diff
@@ -89,13 +90,25 @@ impl PropertyDeclarationBlock {
...
+                            important_count += 1;
+                        }
+                    },
                     None => return Ok(()),
                 }
             }
 
+            // If there is one or more longhand with important, and one or more
+            // without important, we don't serialize it as a shorthand.
```
Comment (emilio) [inline components/style/properties/declaration_block.rs]: Please add the step number (3.3.2.4)

File: components/style/properties/properties.mako.rs [hunk 3]
```diff
@@ -435,7 +435,8 @@ impl Shorthand {
     pub fn serialize_shorthand_to_buffer<'a, W, I>(self,
                                                    dest: &mut W,
                                                    declarations: I,
-                                                   is_first_serialization: &mut bool)
+                                                   is_first_serialization: &mut bool,
+                                                   is_important: bool)
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: Can we pass an `Importance` argument instead? Not a huge deal, but I think it's a bit clearer. Don't bother too much about it if it's too much churn.

**Review by emilio — APPROVED** [review]
r=me with that, thanks!

**Comment by canaltinova** [comment]
Pushed, thanks!

**Pull request merged**